#### Python Defined Function Assignment
#### This notebook contains solutions to Python defined function Assignment questions.
#### ==============================================================

Q.1) Write a Python function to sum all the numbers in a list.
   
Sample List : (8, 2, 3, 0, 7)

Expected Output : 20


In [1]:
# Python function to sum all the numbers in a list

def sum_list(numbers):
    total = 0
    for num in numbers:
        total += num
    return total

# Sample List
sample_list = [8, 2, 3, 0, 7]

# Function Call
print("Sum of the list:", sum_list(sample_list))

Sum of the list: 20


##### Conclusion:This program defines a function sum_list() that calculates the sum of all the numbers in a list using a for loop. Each element is added to the variable total, and the final sum is returned. For the sample list [8, 2, 3, 0, 7], the function returns 20. This approach is simple, efficient, and can be used with any list of numeric values.
#### ==========================================================================

Q.2) Write a Python function to multiply all the numbers in a list. 
    
Sample List : (8, 2, 3, -1, 7)

Expected Output : -336

In [3]:
# Function to multiply all the numbers in a list

def multiply_list(numbers):
    product = 1
    for num in numbers:
        product *= num
    return product

# Sample List
sample_list = [8, 2, 3, -1, 7]

# Function Call
print("Product of the list:", multiply_list(sample_list))

Product of the list: -336


##### Conclusion:The program uses a function and a for loop to multiply all the numbers in a list and returns the final product. For the given list, the output is -336.
#### ==========================================================================

Q.3) Write a Python function to check whether a number is in a given range.

In [4]:
# Function to check whether a number is in a given range

def check_range(num, start, end):
    if num in range(start, end + 1):
        print(num, "is in the range.")
    else:
        print(num, "is not in the range.")

# Function Call
check_range(5, 1, 10)

5 is in the range.


##### Conclusion:The program uses a function to check whether a given number lies within a specified range. If the number is within the range, it displays a confirmation message; otherwise, it indicates that the number is outside the range.
#### ==========================================================================

Q.4) Write a Python function that takes a list and returns a new list with unique elements of the first list.

Sample List : [1,2,3,3,3,3,4,5]

Unique List : [1, 2, 3, 4, 5]


In [5]:
# Function to return unique elements from a list

def unique_list(numbers):
    unique = []
    for num in numbers:
        if num not in unique:
            unique.append(num)
    return unique

# Sample List
sample_list = [1, 2, 3, 3, 3, 3, 4, 5]

# Function Call
print("Unique List:", unique_list(sample_list))

Unique List: [1, 2, 3, 4, 5]


##### Conclusion:The program removes duplicate elements from a list and returns a new list containing only unique values while preserving their original order.
#### ==========================================================================

Q.5) Write a Python function that takes a number as a parameter and check the number is prime or not.

Note : A prime number (or a prime) is a natural number greater than 1 and that has no positive divisors other than 1 and itself.


In [6]:
# Function to check whether a number is prime

def is_prime(num):
    if num <= 1:
        return False

    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            return False
    return True

# Sample Number
number = 17

# Function Call
if is_prime(number):
    print(number, "is a Prime Number.")
else:
    print(number, "is not a Prime Number.")

17 is a Prime Number.


##### Conclusion: The program checks whether a given number is prime by testing if it has any divisors other than 1 and itself. If no divisors are found, it is a prime number; otherwise, it is not.
#### ==========================================================================

Q.6) Write a Python program to print the even numbers from a given list.


In [7]:
# Function to print even numbers from a list

def print_even(numbers):
    print("Even numbers are:")
    for num in numbers:
        if num % 2 == 0:
            print(num)

# Sample List
sample_list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Function Call
print_even(sample_list)

Even numbers are:
2
4
6
8
10


##### Conclusion:The program uses a function and a for loop to check each element in the list and prints only the even numbers using the modulus (%) operator.
#### ==========================================================================

Q.7) Write a Python function to calculate the factorial of a number (a non-negative integer). The function accepts the number as an argument.


In [8]:
# Function to calculate the factorial of a number

def factorial(n):
    if n < 0:
        return "Factorial is not defined for negative numbers."

    fact = 1
    for i in range(1, n + 1):
        fact *= i
    return fact

# Sample Number
number = 5

# Function Call
print("Factorial of", number, "is", factorial(number))

Factorial of 5 is 120


##### Conclusion:The program uses a function and a for loop to calculate the factorial of a non-negative integer. For the given number 5, the factorial is 120.
#### ==========================================================================

Q.8)  Write a Python function to check whether a number is perfect or not. Go to the editor
According to Wikipedia : In number theory, a perfect number is a positive integer that is equal to the sum of its proper positive divisors, that is, the sum of its positive divisors excluding the number itself (also known as its aliquot sum). Equivalently, a perfect number is a number that is half the sum of all of its positive divisors (including itself).

Example : The first perfect number is 6, because 1, 2, and 3 are its proper positive divisors, and 1 + 2 + 3 = 6. Equivalently, the number 6 is equal to half the sum of all its positive divisors: ( 1 + 2 + 3 + 6 ) / 2 = 6. The next perfect number is 28 = 1 + 2 + 4 + 7 + 14. This is followed by the perfect numbers 496 and 8128.


In [9]:
# Function to check whether a number is perfect

def is_perfect(num):
    if num <= 0:
        return False

    total = 0
    for i in range(1, num):
        if num % i == 0:
            total += i

    return total == num

# Sample Number
number = 28

# Function Call
if is_perfect(number):
    print(number, "is a Perfect Number.")
else:
    print(number, "is not a Perfect Number.")

28 is a Perfect Number.


##### Conclusion:The program checks whether a number is equal to the sum of its proper positive divisors. If the sum is equal to the number, it is a perfect number; otherwise, it is not.
#### ==========================================================================

Q.9) Write a Python function to create and print a list where the values are square of numbers between 1 and 30 (both included).


In [10]:
# Function to create and print a list of squares

def square_list():
    squares = []
    for i in range(1, 31):
        squares.append(i ** 2)
    return squares

# Function Call
print(square_list())

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900]


##### Conclusion:The program creates a list containing the squares of numbers from 1 to 30 using a for loop and returns the complete list.
#### ==========================================================================

## <span style="color:purple">Conclusion</span>
In this assignment, I practiced Python user-defined functions and learned how to:

- Create and use functions
- Pass arguments to functions
- Return values from functions
- Find the sum and product of list elements
- Check numbers in a given range
- Remove duplicate elements from a list
- Check prime and perfect numbers
- Print even numbers from a list
- Calculate the factorial of a number
- Generate a list of squares using functions